xalia/Uia/MsaaElement.cs

```diff
@@ -88,7 +121,73 @@ protected override void DumpProperties()
             Utils.DebugWriteLine($"  msaa_pid: {ElementWrapper.Pid}");
             if (!(_processName is null))
                 Utils.DebugWriteLine($"  msaa_process_name: {_processName}");
+            if (_roleKnown)
+            {
+                if (_role >= 0 && _role < UiaElement.msaa_role_to_enum.Length)
+                    Utils.DebugWriteLine($"  msaa_role: {UiaElement.msaa_role_to_enum[_role]}");
+                else
+                    Utils.DebugWriteLine($"  msaa_role: {_role}");
+            }
             base.DumpProperties();
         }
+
+        protected override void WatchProperty(GudlExpression expression)
+        {
+            if (expression is IdentifierExpression id)
+            {
+                switch (id.Name)
+                {
+                    case "msaa_role":
+                        {
+                            if (!_roleKnown && !_fetchingRole)
+                            {
+                                _fetchingRole = true;
+                                Utils.RunTask(FetchRole());
+                            }
+                            break;
+                        }
+                }
+            }
+            base.WatchProperty(expression);
+        }
+
+        private async Task FetchRole()
+        {
+            object role_obj;
+            try
+            {
+                role_obj = await Root.CommandThread.OnBackgroundThread(() =>
+                {
+                    return ElementWrapper.Accessible.accRole[ElementWrapper.ChildId];
+                }, ElementWrapper);
+            }
+            catch (Exception e)
+            {
+                if (!UiaElement.IsExpectedException(e))
+                    throw;
+                return;
+            }
+            if (role_obj is null)
+            {
+                Utils.DebugWriteLine($"WARNING: accRole returned NULL for {this}");
+            }
+            else if (!(role_obj is int role))
+            {
+                Utils.DebugWriteLine($"WARNING: accRole returned {role_obj.GetType()} instead of int for {this}");
+            }
+            else
+            {
+                _role = role;
+                _roleKnown = true;
+                if (MatchesDebugCondition())
+                {
+                    if (_role >= 0 && _role < UiaElement.msaa_role_to_enum.Length)
+                        Utils.DebugWriteLine($"{this}.msaa_role: {UiaElement.msaa_role_to_enum[_role]}");
+                    else
+                        Utils.DebugWriteLine($"{this}.msaa_role: {_role}");

```

Comment:
the:
```
if (_role >= 0 && _role < UiaElement.msaa_role_to_enum.Length)
    UiaElement.msaa_role_to_enum[_role]
else
    _role
```

check is done in a few different places, might be worth trying to create some common code for this perhaps.

Reply:
Will do, thank you.

Reply:
This is fixed in the msaa-element branch, will update the PR when #2 is merged.